staging/src/k8s.io/apimachinery/pkg/util/managedfields/internal/structuredmerge.go

```diff
@@ -44,33 +44,38 @@ func NewStructuredMergeManager(typeConverter TypeConverter, objectConverter runt
 	if typeConverter == nil {
 		return nil, fmt.Errorf("typeconverter must not be nil")
 	}
+	builder := merge.UpdaterBuilder{
```

Comment:
it's hard to tell what this change does... since the PR doesn't change any tests (other than deleting a couple), it's hard to understand what behavior is changing

Reply:
The key difference here is the addition of the new option `ReturnInputOnNoop: true` to `merge.Updater`.  `ReturnInputOnNoop` disables an unnecessary and overly permissive no-op check at the end of the merging to decide if timestamp should be bumped.

Since this code was initially written we have added another check later in the request pipeline to detect no-ops modulo timestamp:

https://github.com/kubernetes/kubernetes/blob/0241da314e0e69817d66313b45a69c19d1ce7327/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L676-L682

That check makes the no-op check at the end of merge unnecessary. The net effect of this PR should be:

1. Improved performance from removed comparison during SSA requests
2. Fixes bug in which re-ordered map-lists and sets would be marked as a no-op (SMD's comparion was aware of structure, the timestamp transformer is not) (correct me if I'm wrong here @apelisse)

Reply:
> Fixes bug in which re-ordered map-lists and sets would be marked as a no-op

If this is fixing that issue, that would be great to have a test for to demonstrate

Reply:
It actually doesn't fix the bug, the bug is already fixed by having bumped the version of smd, it's only addressing 1.
The comparison we do in smd is moot because we re-normalize later (after admission) since a couple of releases ago.
I can add an integration test that shows that comparison is useless, but I sort of assumed that Alex already had a test to check that normalization (via `fieldmanager.IgnoreManagedFieldsTimestampsTransformer`) works.

Reply:
There are tests for resource.Quantity normalization case, so we have confirmation that the comparison is useless:

https://github.com/kubernetes/kubernetes/blob/c2cbc460f2c5be07b0f5762005642625b06b6a75/test/integration/apiserver/apply/apply_test.go#L266

I still think it would be good to add one to confirm the bug was fixed by updating SMD.